src/Parsers/FragmentParser.php

```diff
@@ -12,25 +12,33 @@
 class FragmentParser
 {
 	/**
-	 * @param array $body
+	 * @param array|AbstractFragment $body
 	 *
 	 * @return array
 	 */
-	public function parse(array $body)
+	public function parse($body)
 	{
-		// Have we parsed something, we'll recursively keep parsing until this stays false.
-		$parsed = false;
+		if (is_array($body)) {
+			foreach ($body as $key => &$item) {
+				if ($item instanceof AbstractFragment) {
+					$parsedBody = $this->parse($item->getBody());
 
-		// Replace all abstracts with their body.
-		array_walk_recursive($body, function (&$item) use (&$parsed) {
-			if ($item instanceof AbstractFragment) {
-				$item = $item->getBody();
-				$parsed = true;
+					// Add the parsed fragment to the parent.
+					if ($item->mergeWithParent) {
+						$body = array_merge($body, $parsedBody);
+						unset($body[$key]);
+					} // Replace its current position with the parsed fragment.
+					else {
+						$item = $parsedBody;
+					}
+				} // Further nesting to be parsed.
+				elseif (is_array($item)) {
+					$item = $this->parse($item);
+				}
 			}
-		});
-
-		if ($parsed) {
-			$body = $this->parse($body);
+			// Root level fragment, does not have a parent so it can not be merged with one.
+		} elseif ($body instanceof AbstractFragment) {
+			$body = $this->parse($body->getBody());
```

Comment:
Fragments can be nested, the parser replaces them recursively. These tests test for that:

* https://github.com/madewithlove/elasticsearcher/blob/master/tests/Parsers/FragmentParserTest.php#L63-L95
* https://github.com/madewithlove/elasticsearcher/blob/master/tests/Parsers/FragmentParserTest.php#L63-L95 (just added)